web/packages/teleport/src/AppLauncher/AppLauncher.tsx

```diff
@@ -37,6 +37,43 @@ export function AppLauncher() {
   const queryParams = new URLSearchParams(search);
   const isRedirectFlow = queryParams.get('required-apps');
 
+  function validateResolvedApp(
+    params: UrlLauncherParams,
+    fqdn: string,
+    localClusterName: string,
+    publicAddress: string
+  ) {
+    // Because the ports are stripped from the FQDNs before they are
+    // compared, an attacker can pass a FQDN with a different port than
+    // what the app's public address is configured with and have Teleport
+    // redirect to the public address with an arbitrary port. But because
+    // the attacker can't control what domain is redirected to this has
+    // a low risk factor.
+    if (prepareFqdn(fqdn) == prepareFqdn(params.fqdn)) {
```

Comment:
At the time this check was added in #43525 (e22c004), this check seemed to work correctly, indicating that a regression was introduced elsewhere. Did you look into this by any chance?

FWIW, I couldn't get leaf app access working even on e22c004. Despite the fqdn check passing, I see the following message displayed by `AppLauncher`:

> Unable to serve application requests. Please try again. If the issue persists, verify if the Application Services are connected to Teleport.

The logs for the root cluster show this:

```
2024-10-28T12:15:45+01:00 WARN [PROXY:SER] Request for connection to remote site failed error:[
ERROR REPORT:
Original Error: *errors.errorString connection rejected: could not find reverse tunnel, check that Application Service agent proxying this application is up and running
Stack Trace:
	github.com/gravitational/teleport/api@v0.0.0/utils/sshutils/conn.go:74 github.com/gravitational/teleport/api/utils/sshutils.ConnectProxyTransport
	github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:970 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).chanTransportConn
	github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:943 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).connThroughTunnel
	github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:808 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).DialTCP
	github.com/gravitational/teleport/lib/reversetunnel/remotesite.go:802 github.com/gravitational/teleport/lib/reversetunnel.(*remoteSite).Dial
	github.com/gravitational/teleport/lib/web/app/transport.go:436 github.com/gravitational/teleport/lib/web/app.dialAppServer
	github.com/gravitational/teleport/lib/web/app/match.go:117 github.com/gravitational/teleport/lib/web/app.(*Handler).HealthCheckAppServer.appServerMatcher.MatchHealthy.func2
	github.com/gravitational/teleport/lib/web/app/match.go:131 github.com/gravitational/teleport/lib/web/app.(*Handler).HealthCheckAppServer.appServerMatcher.MatchAll.func3
	github.com/gravitational/teleport/lib/web/app/match.go:76 github.com/gravitational/teleport/lib/web/app.MatchOne
	github.com/gravitational/teleport/lib/web/app/handler.go:236 github.com/gravitational/teleport/lib/web/app.(*Handler).HealthCheckAppServer
	github.com/gravitational/teleport/lib/web/apps.go:223 github.com/gravitational/teleport/lib/web.(*Handler).createAppSession
	github.com/gravitational/teleport/lib/web/apiserver.go:4464 github.com/gravitational/teleport/lib/web.(*Handler).bindDefaultEndpoints.(*Handler).WithAuth.func7
	github.com/gravitational/teleport/lib/httplib/httplib.go:127 github.com/gravitational/teleport/lib/web.(*Handler).bindDefaultEndpoints.(*Handler).WithAuth.MakeHandler.MakeHandlerWithErrorWriter.func161
	github.com/julienschmidt/httprouter@v1.3.0/router.go:399 github.com/julienschmidt/httprouter.(*Router).ServeHTTP
	net/http/server.go:2214 github.com/gravitational/teleport/lib/web.NewHandler.func1.StripPrefix.1
	net/http/server.go:2171 net/http.HandlerFunc.ServeHTTP
	github.com/gravitational/teleport/lib/web/apiserver.go:528 github.com/gravitational/teleport/lib/web.NewHandler.func1
	net/http/server.go:2171 net/http.HandlerFunc.ServeHTTP
	github.com/julienschmidt/httprouter@v1.3.0/router.go:460 github.com/julienschmidt/httprouter.(*Router).ServeHTTP
	github.com/gravitational/teleport/lib/web/apiserver.go:366 github.com/gravitational/teleport/lib/web.(*APIHandler).ServeHTTP
	github.com/gravitational/oxy@v0.0.0-20231219172753-f855322f2a6c/ratelimit/tokenlimiter.go:118 github.com/gravitational/oxy/ratelimit.(*TokenLimiter).ServeHTTP
	github.com/gravitational/oxy@v0.0.0-20231219172753-f855322f2a6c/connlimit/connlimit.go:75 github.com/gravitational/oxy/connlimit.(*ConnLimiter).ServeHTTP
	github.com/gravitational/teleport/lib/httplib/httplib.go:104 github.com/gravitational/teleport/lib/httplib.MakeTracingHandler.func1
	net/http/server.go:2171 net/http.HandlerFunc.ServeHTTP
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.53.0/handler.go:218 go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*middleware).serveHTTP
	go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.53.0/handler.go:74 go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.NewMiddleware.func1.1
	net/http/server.go:2171 net/http.HandlerFunc.ServeHTTP
	net/http/server.go:3142 net/http.serverHandler.ServeHTTP
	net/http/server.go:2044 net/http.(*conn).serve
	runtime/asm_arm64.s:1222 runtime.goexit
User Message: connection rejected: could not find reverse tunnel, check that Application Service agent proxying this application is up and running] cluster:teleport-local-leaf reversetunnel/remotesite.go:947
```

Reply:
> At the time this check was added in https://github.com/gravitational/teleport/pull/43525 (https://github.com/gravitational/teleport/commit/e22c004f5dab6759b681397f1b4441c8573351e7), it seemed to work correctly, indicating that a regression was introduced elsewhere. Did you look into this by any chance?

Yes, see the earlier comment by @avatus. I just tested and pushed another commit based on his diff which should fix this issue.

> FWIW, I couldn't get leaf app access working even on https://github.com/gravitational/teleport/commit/e22c004f5dab6759b681397f1b4441c8573351e7. Despite the fqdn check passing, I see the following message displayed by AppLauncher:

I noticed this behavior too, and @avatus did as well. I chocked it up to a MacOS DNS bug, as my leaf cluster wasn't respecting the entry in my hosts file for the leaf domain. I'm not sure what's causing this or if it's related, but I didn't remember running into the issue back when I pushed the https://github.com/gravitational/teleport/pull/43525.

Re: @avatus's comment on whether the FQDN being changed for a leaf app resource is correct/intended behavior, I'm not sure on that, but I did find that it's not new code. It was introduced in https://github.com/gravitational/teleport/commit/8720b1872a1a9722f6613f5940d422cd35678db5#diff-f635fca1ca174ba271760ef6e8dbe4aa703df1ba9acde1ae8828b86fb8992631 3 years ago and to my knowledge hasn't caused a problem since, I'm sure it's for a reason but am not presently sure what that reason was.

Reply:
@capnspacehook So, I discovered that in e22c004f5dab6 app access wasn't working not because of the code, but because the web PKI cert I had in my leaf cluster had expired. After fixing this and adding appropriate entries in `/etc/hosts`, I got it to work.

After running `git bisect`, I found that it was #45512 that broke app access for leaf apps, which confirms your earlier thoughts.

Reply:
After #45512, accessing a leaf app from a root cluster fails with `Failed to match applications with FQDN "example-leaf.teleport-local-leaf"`. As this check wasn't added in that PR, it suggests that #45512 introduced a regression in `lib/web.Handler.resolveApp`.

Question is, in order to fix this, do we need anything more than adjusting the behavior of `lib/web.Handler.resolveApp`? Are the changes in `AppLauncher` necessary?

Reply:
Ah, it was me all along! Thanks for the find @ravicious .

if you get a fix to `resolveApp` (probably [the conditional I added](https://github.com/gravitational/teleport/pull/45512/files#diff-b14891674f8dd044033299cfe41f2bad8bd692e2661118d685c612523f37f264R348) ) can you make sure that it doesn't break automatic redirects? [I added an easy way to test](https://github.com/gravitational/teleport/blob/master/.github/ISSUE_TEMPLATE/webtestplan.md#required-applications) it in the web test plan that you can use to check.

Reply:
Ok, i nailed it down to what caused the regression. During required app redirects, I changed [the hostname to the publicaddr of the redirect app](https://github.com/gravitational/teleport/pull/45512/files#diff-00b478b3d7d6aa201917b8129d4a7f923c5e268854c0f1776ab3b757341b0752R84) instead of the host in the request. 

Reverting this change fixes apps on leafs but then breaks required app redirects (causes an infinite loop). I reverted the change in the diff below and then moved the logic to change the hostname based on app redirect only instead of always.

This makes both scenarios work, automatic redirects for required apps AND also apps on leaf. I **did not** get a chance to test required apps _on leafs_. 

Also,I can confirm that [apps of the same name on root and leaf ](https://github.com/gravitational/teleport/pull/47778#discussion_r1819951083) causes the user to be redirected to the app on root (even if selecting the leaf). but i think that is a separate issue not related to any of these suggested changes or regression (but what do i know).

<details><summary>Open me!</summary>

```diff
diff --git a/lib/web/app/handler.go b/lib/web/app/handler.go
index 38de429b9c..e81f686042 100644
--- a/lib/web/app/handler.go
+++ b/lib/web/app/handler.go
@@ -618,10 +618,14 @@ const (
 // The URL's are formed this way to help isolate the path params reserved for the app
 // launchers route, where order and existence of previous params matter for this route.
 func makeAppRedirectURL(r *http.Request, proxyPublicAddr, hostname string, req launcherURLParams) string {
+	host := hostname
+	if req.requiresAppRedirect {
+		host = req.publicAddr
+	}
 	u := url.URL{
 		Scheme: "https",
 		Host:   proxyPublicAddr,
-		Path:   fmt.Sprintf("/web/launch/%s", hostname),
+		Path:   fmt.Sprintf("/web/launch/%s", host),
 	}
 
 	// Presence of a stateToken means we are beginning an app auth exchange.
@@ -634,7 +638,7 @@ func makeAppRedirectURL(r *http.Request, proxyPublicAddr, hostname string, req l
 		v.Add("required-apps", req.requiredAppFQDNs)
 		u.RawQuery = v.Encode()
 
-		urlPath := []string{"web", "launch", hostname}
+		urlPath := []string{"web", "launch", host}
 
 		// The order and existence of previous params matter.
 		//
diff --git a/lib/web/app/middleware.go b/lib/web/app/middleware.go
index 79e1586ff2..67b189de3a 100644
--- a/lib/web/app/middleware.go
+++ b/lib/web/app/middleware.go
@@ -81,12 +81,8 @@ func (h *Handler) redirectToLauncher(w http.ResponseWriter, r *http.Request, p l
 			"https://goteleport.com/docs/application-access/guides/connecting-apps/#start-authproxy-service.")
 		return trace.BadParameter("public address of the proxy is not set")
 	}
-	host := p.publicAddr
-	if host == "" {
-		host = r.Host
-	}
 
-	addr, err := utils.ParseAddr(host)
+	addr, err := utils.ParseAddr(r.Host)
 	if err != nil {
 		return trace.Wrap(err)
 	}

```
</details>

this was tested off of master btw, so unless there are some _other_ regressions or security holes filled by your first commit, your og change may not be necessary. 

Please take time to test this as I did some light testing locally but may have missed an edge case.